src/gui/processing/qgsprocessingtoolboxmodel.cpp

```diff
@@ -0,0 +1,805 @@
+/***************************************************************************
+    QgsProcessingToolboxModel.cpp
+    -------------------------------
+    begin                : May 2018
+    copyright            : (C) 2018 by Nyall Dawso
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsprocessingtoolboxmodel.h"
+#include "qgsapplication.h"
+#include "qgsprocessingregistry.h"
+#include "qgsprocessingrecentalgorithmlog.h"
+#include <functional>
+
+#ifdef ENABLE_MODELTEST
+#include "modeltest.h"
+#endif
+
+///@cond PRIVATE
+
+//
+// QgsProcessingToolboxModelNode
+//
+
+QgsProcessingToolboxModelNode::~QgsProcessingToolboxModelNode()
+{
+  deleteChildren();
+}
+
+QgsProcessingToolboxModelNode *QgsProcessingToolboxModelNode::takeChild( QgsProcessingToolboxModelNode *node )
+{
+  return mChildren.takeAt( mChildren.indexOf( node ) );
+}
+
+QgsProcessingToolboxModelGroupNode *QgsProcessingToolboxModelNode::getChildGroupNode( const QString &groupId )
+{
+  for ( QgsProcessingToolboxModelNode *node : qgis::as_const( mChildren ) )
+  {
+    if ( node->nodeType() == NodeGroup )
+    {
+      QgsProcessingToolboxModelGroupNode *groupNode = qobject_cast< QgsProcessingToolboxModelGroupNode * >( node );
+      if ( groupNode && groupNode->id() == groupId )
+        return groupNode;
+    }
+  }
+  return nullptr;
+}
+
+void QgsProcessingToolboxModelNode::addChildNode( QgsProcessingToolboxModelNode *node )
+{
+  if ( !node )
+    return;
+
+  Q_ASSERT( !node->mParent );
+  node->mParent = this;
+
+  mChildren.append( node );
+}
+
+void QgsProcessingToolboxModelNode::deleteChildren()
+{
+  qDeleteAll( mChildren );
+  mChildren.clear();
+}
+
+//
+// QgsProcessingToolboxModelProviderNode
+//
+
+QgsProcessingToolboxModelProviderNode::QgsProcessingToolboxModelProviderNode( QgsProcessingProvider *provider )
+  : mProviderId( provider->id() )
```

Comment:
Is it necessary to keep the provider id as a member (considering that there's `mProvider` already)? Moreover, provider may be a `nullptr`.

Reply:
Actually this is necessary, because it's used to identify which provider the node is linked to for cleanups after a provider is removed from the registry. But I've added some comments explaining the logic here, and there's unit tests in place protecting this.